compiler_opt/rl/generate_bc_trajectories.py

```diff
@@ -529,3 +571,248 @@ def _process_obs(self, curr_obs, sequence_example):
+    self._exploration_policy_paths: Optional[str] = exploration_policy_paths
+    self._exploration_frac: float = exploration_frac
+    self._max_exploration_steps: int = max_exploration_steps
+    self._tf_policy_action: List[Optional[Callable[[Any], np.ndarray]]] = []
+    self._exploration_policy_distrs: Optional[List[
+        Callable[[time_step.TimeStep, Optional[tf_types.NestedTensor]],
+                 policy_step.PolicyStep]]] = None
+    self._explore_on_features: Optional[Dict[str, Callable[
+        [tf.Tensor], bool]]] = explore_on_features
+    self._obs_action_specs: Optional[Tuple[
+        time_step.TimeStep, tensor_spec.BoundedTensorSpec]] = obs_action_specs
+    self._mw_utility = ModuleWorkerUtility(base_path)
+    self._partitions = partitions
+    self._envargs = envargs
+
+    for policy_path in policy_paths:
+      tf_policy = tf.saved_model.load(policy_path, tags=None, options=None)
+      self._tf_policy_action.append(policy_action_wrapper(tf_policy))
+    if exploration_policy_paths:
+      if len(exploration_policy_paths) != len(policy_paths):
+        raise AssertionError(
+            f'Number of exploration policies: {0}, does not match number of policies: {1}'  # pylint: disable=line-too-long
+            .format(len(exploration_policy_paths), len(policy_paths)))
+      self._exploration_policy_distrs = []
+      for exploration_policy_path in exploration_policy_paths:
+        expl_policy = tf.saved_model.load(
+            exploration_policy_path, tags=None, options=None)
+        self._exploration_policy_distrs.append(
+            policy_distr_wrapper(expl_policy))
+
+  def select_best_exploration(
+      self,
+      loaded_module_spec: corpus.LoadedModuleSpec,
+  ) -> tf.train.SequenceExample:
+
+    num_calls = len(self._tf_policy_action)
+    time_call_compiler = 0
+    logging.info('Processing module: %s', loaded_module_spec.name)
+    start = timeit.default_timer()
+    work = list(zip(self._tf_policy_action, self._exploration_policy_distrs))
+    exploration_worker = ExplorationWorker(
+        loaded_module_spec=loaded_module_spec,
+        clang_path=self._clang_path,
+        mlgo_task=self._mlgo_task,
+        exploration_frac=self._exploration_frac,
+        max_exploration_steps=self._max_exploration_steps,
+        explore_on_features=self._explore_on_features,
+        obs_action_specs=self._obs_action_specs,
+        **self._envargs)
+    succeeded = []
+    for policy_action, explore_policy in work:
+      exploration_res = None
+      try:
+        exploration_res = exploration_worker.explore_function(
+            policy_action, explore_policy)
+      except Exception as e:  # pylint: disable=broad-except
+        logging.info('Compilation exception %s at %s', e,
+                     loaded_module_spec.name)
+      if exploration_res is not None:
+        succeeded.append(exploration_res)
+
+    end = timeit.default_timer()
+    time_call_compiler += end - start
```

Comment:
Is there a reason we need to be recording this here? It seems like something intended for debugging that should probably be pulled out.

Reply:
When collecting a dataset I often found it useful to keep track of how long a module is taking to compile as this gives me a rough idea of how long the whole corpus compilation is going to take. I can of course remove this but will likely end up adding it again for personal use the next time we want to train a new model with BC.

Reply:
Ack.

Could we pass this around in a different manner? Maybe just a `logging.info` call after computing the time taken? I can't see how you're planning on using it, but I think it would be a bit nicer if we can avoid passing this around.

Reply:
Oh, yeah definitely. I think I was using it exactly for logging in the main function which creates a ModuleWorker and process the results from that but it's probably best to have it here.